File: neumann-py/src/neumann/retry.py
# SPDX-License-Identifier: MIT
"""Retry logic with exponential backoff for transient failures."""

from __future__ import annotations

import asyncio
import random
import time
from functools import wraps
from typing import TYPE_CHECKING, Awaitable, Callable, ParamSpec, TypeVar

if TYPE_CHECKING:
    from neumann.config import RetryConfig

T = TypeVar("T")
P = ParamSpec("P")


def _is_retryable(e: Exception, config: RetryConfig) -> bool:
    """Check if exception is retryable based on gRPC status code."""
    if hasattr(e, "code") and callable(e.code):
        try:
            code = e.code()
            if hasattr(code, "value") and isinstance(code.value, tuple):
                return code.value[0] in config.retryable_status_codes
        except Exception:
            pass
    return False


def _calculate_backoff(
    attempt: int,
    initial_ms: int,
    max_ms: int,
    multiplier: float,
) -> float:
    """Calculate backoff time with jitter."""
    backoff_ms = initial_ms * (multiplier**attempt)
    jitter = random.uniform(0.8, 1.2)
    return min(backoff_ms * jitter, max_ms)


def with_retry(config: RetryConfig) -> Callable[[Callable[P, T]], Callable[P, T]]:
    """Decorator for sync functions with retry and exponential backoff."""

    def decorator(func: Callable[P, T]) -> Callable[P, T]:
        @wraps(func)
        def wrapper(*args: P.args, **kwargs: P.kwargs) -> T:
            last_exception: Exception | None = None

            for attempt in range(config.max_attempts):
                try:
                    return func(*args, **kwargs)
                except Exception as e:
                    if not _is_retryable(e, config):
                        raise
                    last_exception = e
                    if attempt < config.max_attempts - 1:
                        sleep_ms = _calculate_backoff(
                            attempt,
                            config.initial_backoff_ms,
                            config.max_backoff_ms,
                            config.backoff_multiplier,
                        )
                        time.sleep(sleep_ms / 1000)

            if last_exception is not None:
                raise last_exception
            raise RuntimeError("Retry loop completed without result")

        return wrapper

    return decorator


def with_retry_async(
    config: RetryConfig,
) -> Callable[[Callable[P, Awaitable[T]]], Callable[P, Awaitable[T]]]:
    """Decorator for async functions with retry and exponential backoff."""

    def decorator(func: Callable[P, Awaitable[T]]) -> Callable[P, Awaitable[T]]:
        @wraps(func)
        async def wrapper(*args: P.args, **kwargs: P.kwargs) -> T:
            last_exception: Exception | None = None

            for attempt in range(config.max_attempts):
                try:
                    return await func(*args, **kwargs)
                except Exception as e:
                    if not _is_retryable(e, config):
                        raise
                    last_exception = e
                    if attempt < config.max_attempts - 1:
                        sleep_ms = _calculate_backoff(
                            attempt,
                            config.initial_backoff_ms,
                            config.max_backoff_ms,
                            config.backoff_multiplier,
                        )
                        await asyncio.sleep(sleep_ms / 1000)

            if last_exception is not None:
                raise last_exception
            raise RuntimeError("Retry loop completed without result")

        return wrapper

    return decorator


def retry_call(
    func: Callable[[], T],
    config: RetryConfig,
) -> T:
    """Execute a function with retry logic.

    This is a non-decorator approach for cases where you need to retry
    a specific call inline.

    Args:
        func: Zero-argument callable to retry.
        config: Retry configuration.

    Returns:
        Result of the function call.

    Raises:
        The last exception if all retries are exhausted.
    """
    last_exception: Exception | None = None

    for attempt in range(config.max_attempts):
        try:
            return func()
        except Exception as e:
            if not _is_retryable(e, config):
                raise
            last_exception = e
            if attempt < config.max_attempts - 1:
                sleep_ms = _calculate_backoff(
                    attempt,
                    config.initial_backoff_ms,
                    config.max_backoff_ms,
                    config.backoff_multiplier,
                )
                time.sleep(sleep_ms / 1000)

    if last_exception is not None:
        raise last_exception
    raise RuntimeError("Retry loop completed without result")


async def retry_call_async(
    func: Callable[[], Awaitable[T]],
    config: RetryConfig,
) -> T:
    """Execute an async function with retry logic.

    Args:
        func: Zero-argument async callable to retry.
        config: Retry configuration.

    Returns:
        Result of the function call.

    Raises:
        The last exception if all retries are exhausted.
    """
    last_exception: Exception | None = None

    for attempt in range(config.max_attempts):
        try:
            return await func()
        except Exception as e:
            if not _is_retryable(e, config):
                raise
            last_exception = e
            if attempt < config.max_attempts - 1:
                sleep_ms = _calculate_backoff(
                    attempt,
                    config.initial_backoff_ms,
                    config.max_backoff_ms,
                    config.backoff_multiplier,
                )
                await asyncio.sleep(sleep_ms / 1000)

    if last_exception is not None:
        raise last_exception
    raise RuntimeError("Retry loop completed without result")

File: neumann-ts/src/retry.ts
// SPDX-License-Identifier: MIT
/**
 * Retry logic with exponential backoff for transient failures.
 */

/**
 * gRPC status codes that are typically retryable.
 */
export const RETRYABLE_STATUS_CODES = {
  DEADLINE_EXCEEDED: 4,
  UNAVAILABLE: 14,
} as const;

/**
 * Configuration for retry behavior.
 */
export interface RetryConfig {
  /** Maximum number of retry attempts (default: 3). */
  maxAttempts: number;
  /** Initial backoff duration in milliseconds (default: 100). */
  initialBackoffMs: number;
  /** Maximum backoff duration in milliseconds (default: 10000). */
  maxBackoffMs: number;
  /** Multiplier for exponential backoff (default: 2.0). */
  backoffMultiplier: number;
  /** gRPC status codes that should trigger a retry. */
  retryableStatusCodes: number[];
}

/**
 * Default retry configuration.
 */
export const DEFAULT_RETRY_CONFIG: RetryConfig = {
  maxAttempts: 3,
  initialBackoffMs: 100,
  maxBackoffMs: 10000,
  backoffMultiplier: 2.0,
  retryableStatusCodes: [
    RETRYABLE_STATUS_CODES.UNAVAILABLE,
    RETRYABLE_STATUS_CODES.DEADLINE_EXCEEDED,
  ],
};

/**
 * Partial retry configuration for overrides.
 */
export type PartialRetryConfig = Partial<RetryConfig>;

/**
 * Merge partial config with defaults.
 */
export function mergeRetryConfig(partial: PartialRetryConfig): RetryConfig {
  return { ...DEFAULT_RETRY_CONFIG, ...partial };
}

/**
 * Check if an error is retryable based on gRPC status code.
 */
export function isRetryable(error: unknown, config: RetryConfig): boolean {
  if (error === null || typeof error !== 'object') {
    return false;
  }

  const e = error as { code?: number };
  if (typeof e.code !== 'number') {
    return false;
  }

  return config.retryableStatusCodes.includes(e.code);
}

/**
 * Calculate backoff time with jitter.
 *
 * @param attempt - The current attempt number (0-indexed).
 * @param initialMs - Initial backoff in milliseconds.
 * @param maxMs - Maximum backoff in milliseconds.
 * @param multiplier - Exponential multiplier.
 * @returns Backoff time in milliseconds.
 */
export function calculateBackoff(
  attempt: number,
  initialMs: number,
  maxMs: number,
  multiplier: number
): number {
  const backoffMs = initialMs * Math.pow(multiplier, attempt);
  // Add jitter between 0.8 and 1.2
  const jitter = 0.8 + Math.random() * 0.4;
  return Math.min(backoffMs * jitter, maxMs);
}

/**
 * Sleep for a given number of milliseconds.
 */
function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

/**
 * Execute an async function with retry logic.
 *
 * @param fn - The async function to execute.
 * @param config - Retry configuration.
 * @returns The result of the function.
 * @throws The last error if all retries are exhausted.
 */
export async function withRetry<T>(
  fn: () => Promise<T>,
  config: RetryConfig = DEFAULT_RETRY_CONFIG
): Promise<T> {
  let lastError: unknown = null;

  for (let attempt = 0; attempt < config.maxAttempts; attempt++) {
    try {
      return await fn();
    } catch (error) {
      if (!isRetryable(error, config)) {
        throw error;
      }

      lastError = error;

      if (attempt < config.maxAttempts - 1) {
        const backoffMs = calculateBackoff(
          attempt,
          config.initialBackoffMs,
          config.maxBackoffMs,
          config.backoffMultiplier
        );
        await sleep(backoffMs);
      }
    }
  }

  if (lastError !== null) {
    throw lastError;
  }

  throw new Error('Retry loop completed without result');
}

/**
 * Create a retryable wrapper for an async function.
 *
 * @param fn - The async function to wrap.
 * @param config - Retry configuration.
 * @returns A wrapped function that will retry on transient failures.
 */
export function withRetryWrapper<TArgs extends unknown[], TResult>(
  fn: (...args: TArgs) => Promise<TResult>,
  config: RetryConfig = DEFAULT_RETRY_CONFIG
): (...args: TArgs) => Promise<TResult> {
  return (...args: TArgs) => withRetry(() => fn(...args), config);
}
